src/pow.cpp

```diff
@@ -12,16 +12,74 @@
 
 unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader *pblock, const Consensus::Params& params)
 {
-    assert(pindexLast != nullptr);
-    unsigned int nProofOfWorkLimit = UintToArith256(params.powLimit).GetCompact();
 
-    int nHeightNext = pindexLast->nHeight + 1;
-    if (nHeightNext >= params.BTGHeight  && nHeightNext < params.BTGHeight + params.BTGPremineWindow)
-    {
-        // Lowest difficulty for Bitcoin GPU premining period.
+    
+    unsigned int nProofOfWorkLimit = UintToArith256(params.powLimit).GetCompact();
+    
+    // Genesis block
+    if (pindexLast == NULL)
         return nProofOfWorkLimit;
+    
+    if (pindexLast->nHeight < params.BTGHeight) {
+        return BitcoinGetNextWorkRequired(pindexLast, pblock, params);
     }
-    else if (nHeightNext % params.DifficultyAdjustmentInterval() != 0)
+    else if (pindexLast->nHeight < params.BTGHeight + params.BTGPremineWindow) {
+        return nProofOfWorkLimit;
+    }
+    
+    const CBlockIndex* pindexFirst = pindexLast;
+    arith_uint256 bnTot {0};
+    for (int i = 0; pindexFirst && i < params.nPowAveragingWindow; i++) {
+        arith_uint256 bnTmp;
+        bnTmp.SetCompact(pindexFirst->nBits);
+        bnTot += bnTmp;
+        pindexFirst = pindexFirst->pprev;
+    }
+    
+    if (pindexFirst == NULL)
+        return nProofOfWorkLimit;
+    
+    arith_uint256 bnAvg {bnTot / params.nPowAveragingWindow};
+    
+
+    return CalculateNextWorkRequired(bnAvg, pindexLast->GetMedianTimePast(), pindexFirst->GetMedianTimePast(), params);
+}
+
+unsigned int CalculateNextWorkRequired(arith_uint256 bnAvg, int64_t nLastBlockTime, int64_t nFirstBlockTime, const Consensus::Params& params)
+{
+    
+    // Limit adjustment
+    int64_t nActualTimespan = nLastBlockTime - nFirstBlockTime;
+    nActualTimespan = params.AveragingWindowTimespan() + (nActualTimespan - params.AveragingWindowTimespan())/4;
+    
+    if (nActualTimespan < params.MinActualTimespan())
+        nActualTimespan = params.MinActualTimespan();
+    if (nActualTimespan > params.MaxActualTimespan())
+        nActualTimespan = params.MaxActualTimespan();
+
+    // Retarget
+    const arith_uint256 bnPowLimit = UintToArith256(params.powLimit);
+    arith_uint256 bnNew {bnAvg};
+    bnNew /= params.AveragingWindowTimespan();
+    bnNew *= nActualTimespan;
+    
+    if (bnNew > bnPowLimit)
+        bnNew = bnPowLimit;
+
+    return bnNew.GetCompact();
+}
+
+
+// Depricated for Bitcoin Gold
+unsigned int BitcoinGetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader *pblock, const Consensus::Params& params)
+{
+    assert(pindexLast != nullptr);
+    unsigned int nProofOfWorkLimit = UintToArith256(params.powLimit).GetCompact();
+    
+    int nHeightNext = pindexLast->nHeight + 1;
+    int diffAdjustmentInterval = 2016; // every 2016 blocks for Bitcoin
```

Comment:
It’s generally not a good idea to hardcode the interval here. However, feel free to leave a TODO here and let’s move on.